pkg/machines/helpers.js

```diff
@@ -375,3 +377,164 @@ export function timeoutedPromise(promise, delay, afterTimeoutHandler, afterTimeo
+    case "pci": {
+        // convert hexadecimal number in string to decimal number in string
+        const domain = parseInt(hostdev.source.address.domain, 16).toString();
+        const bus = parseInt(hostdev.source.address.bus, 16).toString();
+        const slot = parseInt(hostdev.source.address.slot, 16).toString();
+        const func = parseInt(hostdev.source.address.func, 16).toString();
+        nodeDev = nodeDevices.find(d => {
+            if ((domain && bus && slot && func) &&
+                d.capability.domain &&
+                d.capability.bus &&
+                d.capability.slot &&
+                d.capability.function &&
+                d.capability.domain._value == domain &&
+                d.capability.bus._value == bus &&
+                d.capability.slot._value == slot &&
+                d.capability.function._value == func)
+                return true;
+        });
+        break;
+    }
+    case "scsi": {
+        const bus = hostdev.source.address.bus;
+        const target = hostdev.source.address.target;
+        const unit = hostdev.source.address.unit;
+        nodeDev = nodeDevices.find(d => {
+            if ((bus && target && unit) &&
+                d.capability.bus &&
+                d.capability.lun &&
+                d.capability.target &&
+                d.capability.bus._value == bus &&
+                d.capability.lun._value == unit &&
+                d.capability.target._value == target)
+                return true;
+        });
+        break;
+    }
+    case "scsi_host": {
+        // TODO add scsi_host
+        nodeDev = undefined;
+        break;
+    }
+    case "mdev": {
+        const uuid = hostdev.source.address.uuid;
+        nodeDev = nodeDevices.find(d => {
+            if (d.path &&
+                d.path._value.contains(uuid))
+                return true;
+        });
+        break;
+    }
+    }
+    return nodeDev;
+}
+
+export function getBootOrderDevices(vm) {
+    let devices = [];
+    let diskCount = 0;
+    let ifaceCount = 0;
+
+    // Some disks and interfaces may have boot order in os->boot (legacy)
+    if (vm.osBoot) {
+        for (let i = 0; i < vm.osBoot.length; i++) {
+            const boot = vm.osBoot[i];
+
+            if (boot.type === "disk" | boot.type === "cdrom" | boot.type === "fd") {
+                devices.push({
+                    device: Object.values(vm.disks)[diskCount++],
+                    bootOrder: i + 1, // bootOrder begins at 1
+                    type: "disk"
+                });
+            } else if (boot.type === "network") {
+                devices.push({
+                    device: Object.values(vm.interfaces)[ifaceCount++],
+                    bootOrder: i + 1, // bootOrder begins at 1
+                    type: "network"
+                });
+            }
+        }
+    }
+
+    const disks = Object.values(vm.disks);
+    // we ignore the disks which have been loaded from os->boot, and load the rest of disks
```

Comment:
> he boot element can be repeated multiple times to setup a priority list of boot devices to try in turn
Nope, I also tried it and you can actually add multiple disks in boot->order.

Hmm, but another thing which I have noticed now, if you add N number of disks to boot->order, libvirt doesn't check if VM actually has N disks defined in devices. So I need to check that, otherwise it can crash.

Reply:
> `boot element can be repeated multiple times to setup a priority list of boot devices to try in turn`
> 

You can repeat it multiple times but for different device types. Like
``` xml
  <boot dev='hd'/>
  <boot dev='cdrom'/>
```

> Nope, I also tried it and you can actually add multiple disks in boot->order.

You can add multiple disks but of different type aka file, cdrom.
But we don't take that in mind. See comment above. 

> 
> Hmm, but another thing which I have noticed now, if you add N number of disks to boot->order, libvirt doesn't check if VM actually has N disks defined in devices. So I need to check that, otherwise it can lead to error.

Good point. Libvirt should check that I guess though.

Reply:
I did not check the code so I don't know what you are planning to use, but please don't use the old boot order configuration.  Cockpit should be able to understand and parse the old <boot> element in the <os> element but it should never create such XML and it should always use the new way of per-device <boot> element.

The old way has a lot of issues especially if domain has a lot of disks and they are using different bus.

Reply:
@Antique Exactly, right now boot order in this PR can read both old way and new way, but when you configure and save, it will always generate "new way XML".

Problem I mentioned was that if cockpit reads old boot order (os->boot) which contains let's say 4 disks, but VM has only 1 disk defined, there will be internal error in cockpit, since libvirt doesn't check that. (I think it will create array of disks with 1 defined and 3 undefined values or something like that). But I will fix that.

Reply:
> You can repeat it multiple times but for different device types. Like
> 
> ```
>   <boot dev='hd'/>
>   <boot dev='cdrom'/>
> ```
Weird, I just configure XML of my VM with this and it works fine:
```
    <boot dev='hd'/>
    <boot dev='hd'/>
    <boot dev='network'/>
    <boot dev='network'/>
```